src/devices/bus/msx_cart/disk.cpp

```diff
@@ -487,31 +663,686 @@ u8 msx_cart_disk_type2_device::status_r()
+	io_space().install_read_handler(0xd3, 0xd3, read8smo_delegate(*m_fdc, FUNC(wd_fdc_analog_device_base::data_r)));
+	io_space().install_read_handler(0xd4, 0xd4, read8smo_delegate(*this, FUNC(msx_cart_avdpf550_device::status_r)));
+}
+
+void msx_cart_avdpf550_device::device_post_load()
+{
+	control_w(m_control);
+}
+
+void msx_cart_avdpf550_device::control_w(u8 control)
+{
+	m_control = control;
+
+	switch (m_control & 0x03)
+	{
+	case 0x01:
+		m_floppy = m_floppy0 ? m_floppy0->get_device() : nullptr;
+		break;
+
+	case 0x02:
+		m_floppy = m_floppy1 ? m_floppy1->get_device() : nullptr;
+		break;
+
+	default:
+		m_floppy = nullptr;
+		break;
+	}
+
+	if (m_floppy)
+	{
+		m_floppy->mon_w(0);
+		m_floppy->ss_w(BIT(m_control, 2) ? 1 : 0);
+	}
+
+	m_fdc->set_floppy(m_floppy);
+}
+
+u8 msx_cart_avdpf550_device::status_r()
+{
+	return 0x3f | (m_fdc->drq_r() ? 0 : 0x40) | (m_fdc->intrq_r() ? 0x80 : 0);
+}
+
+
+DEFINE_DEVICE_TYPE(MSX_CART_AVDPF550, msx_cart_avdpf550_device, "msx_cart_avdpf550", "MSX Cartridge - AVT DPF-550")
+DEFINE_DEVICE_TYPE(MSX_CART_CDX2,     msx_cart_cdx2_device,     "msx_cart_cdx2",     "MSX Cartridge - CDX-2")
+DEFINE_DEVICE_TYPE(MSX_CART_DDX3,     msx_cart_ddx3_device,     "msx_cart_ddx3",     "MSX Cartridge - DDX3.0")
+DEFINE_DEVICE_TYPE(MSX_CART_FD03,     msx_cart_fd03_device,     "msx_cart_fd03",     "MSX Cartridge - FD-03")
+DEFINE_DEVICE_TYPE(MSX_CART_FD051,    msx_cart_fd051_device,    "msx_cart_fd051",    "MSX Cartridge - FD-051")
+DEFINE_DEVICE_TYPE(MSX_CART_FSFD1,    msx_cart_fsfd1_device,    "msx_cart_fsfd1",    "MSX Cartridge - FS-FD1")
+DEFINE_DEVICE_TYPE(MSX_CART_FSFD1A,   msx_cart_fsfd1a_device,   "msx_cart_fsfd1a",   "MSX Cartridge - FS-FD1A")
+DEFINE_DEVICE_TYPE(MSX_CART_FSCF351,  msx_cart_fscf351_device,  "msx_cart_fscf351",  "MSX Cartridge - FS-CF351")
+DEFINE_DEVICE_TYPE(MSX_CART_HB3600,   msx_cart_hb3600_device,   "msx_cart_hb3600",   "MSX Cartridge - HB-3600")
+DEFINE_DEVICE_TYPE(MSX_CART_HBD20W,   msx_cart_hbd20w_device,   "msx_cart_hbd20w",   "MSX Cartridge - HBD-20W")
+DEFINE_DEVICE_TYPE(MSX_CART_HBD50,    msx_cart_hbd50_device,    "msx_cart_hbd50",    "MSX Cartridge - HBD-50")
+DEFINE_DEVICE_TYPE(MSX_CART_HBDF1,    msx_cart_hbdf1_device,    "msx_cart_hbdf1",    "MSX Cartridge - HBD-F1")
+DEFINE_DEVICE_TYPE(MSX_CART_HXF101PE, msx_cart_hxf101pe_device, "msx_cart_hxf101pe", "MSX Cartridge - HX-F101PE")
+DEFINE_DEVICE_TYPE(MSX_CART_MFD001,   msx_cart_mfd001_device,   "msx_cart_mfd001",   "MSX Cartridge - MFD-001")
+DEFINE_DEVICE_TYPE(MSX_CART_ML30DC,   msx_cart_ml30dc_device,   "msx_cart_ml30dc",   "MSX Cartridge - ML-30DC")
+DEFINE_DEVICE_TYPE(MSX_CART_NMS1200,  msx_cart_nms1200_device,  "msx_cart_nms1200",  "MSX Cartridge - NMS-1200")
+DEFINE_DEVICE_TYPE(MSX_CART_TADPF550, msx_cart_tadpf550_device, "msx_cart_tadpf550", "MSX Cartridge - Talent DPF-550")
+DEFINE_DEVICE_TYPE(MSX_CART_VY0010,   msx_cart_vy0010_device,   "msx_cart_vy0010",   "MSX Cartridge - VY0010")
```

Comment:
Since the concrete device classes are not exposed, you can put the device classes in an anonymous namespace to keep exported symbols down.

Reply:
How can I do that? DECLARE_DEVICE_TYPE and DEFINE_DEVICE_TYPE want to be declared at global scope only.

Reply:
You can look at e.g. Game Boy or Apple II cards that are like that.

Reply:
> You can look at e.g. Game Boy or Apple II cards that are like that.

Aha, so DEFINE_DEVICE_TYPE_PRIVATE combined with declaring everything as the interface is the trick. Thanks